Add test scenarios to PR template

AvantikaDG opened the pull request:
Add test scenarios to PR template to include different test cases. This will enforce high quality commits.

Task: BABEL-3626
Signed-off-by: Avantika Dasgupta <davantik@amazon.com>

AvantikaDG committed:
Add test scenarios to PR template

Deepesh125 committed:
Cherry picking few community changes related to SSL/TLS handling into babelfishpg_tds extension (#695)

This commit aims to add few enhancements in SSL/TLS handling code path. For example, better handling of read or write failure inside SslHandShakeRead
and SslHandShakeWrite APIs.

Also, This commit cherry picks following list of commits from community to babelfishpg_tds extension:
1. https://github.com/postgres/postgres/commit/e0e569e1d192c3fed942257302f24b550cf982f4
2. https://github.com/postgres/postgres/commit/4e89c79a52f8a898edd648b56a00f0f4f840cfe7
3. https://github.com/postgres/postgres/commit/79dfa8afb296e1b0dcffbe674e04c5f25cc13dfd
4. https://github.com/postgres/postgres/commit/3fa44a30049826bfe2fd58eec0e8acabd5757411
5. https://github.com/postgres/postgres/commit/f7cd5896a69621818189fbdd209fb2e1fc008102
6. https://github.com/postgres/postgres/commit/e1cc25f59a8a90d821aaf894e1691575ed94454e
7. https://github.com/postgres/postgres/commit/52eec1c53aa6a7df1683fba79078793f1d0eba42
8. https://github.com/postgres/postgres/commit/af0e79c8f4f4c3c2306855045c0d02a6be6485f0
9. https://github.com/postgres/postgres/commit/f5465fade90827534fbd0b795d18dc62e56939e9
10. https://github.com/postgres/postgres/commit/4b12ab18c9d0735d760bf7667b158707b06e5df8
11. https://github.com/postgres/postgres/commit/e835e89a0fd267871e7fbddc39ad00ee3d0cb55c
12. https://github.com/postgres/postgres/commit/6d7a6feac48b1970c4cd127ee65d4c487acbb5e9
13. https://github.com/postgres/postgres/commit/01e6f1a842f406170e5f717305e4a6cf0e84b3ee
14. https://github.com/postgres/postgres/commit/b88377ad65dfa714e1bd1c041421446065c07c9c

Task: BABEL-3574
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

shalinilohia50 committed:
Support GRANT/REVOKE CONNECT TO/FROM user (#677)

* Support GRANT/REVOKE CONNECT TO/FROM user

1. Create guest user on any user created database by default and disable the guest user.
2. Support enabling/disabling of any user on user created databases.
3. Added a column to the catalog table "sys.babelfish_authid_user_ext".
4. The column "user_can_connect" will track if any user has the connect permission to a database.
5. The value of "user_can_connect" will be 1 if the user has permission
   and on a database otherwise 0.
6. Upgrade Scenario: When the database is created on an old instance and
   after the instance is upgraded to the Narwhal, it should create guest
   user on all the database and disable it by default.

Task: BABEL-2678
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

* Remove repeated lines from the upgrade script

* Fix indentation

* Update a function

* simplify function

* Rename the update_guest_catalog function

* fix test

* User 'guest' should not be allowed to be dropped in any database

* Add tests

* Add tests for grant/revoke to other user

* set the migration mode before running the upgrade scripts

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

rishabhtanwar29 committed:
Mitigate upgrade issue due to DROP OPERATOR-CLASS/VIEW (#672)

Description
Previously, we were dropping fixeddecimal_ops operator class
so that we can re-create it in operator family fixeddecimal_ops
but it is not required since PG automatically creates operator family
internally with same name for each operator class.
So, this commit removes redundant operation of DROP/RECREATE
of fixeddecimal_ops operator class/family.
Similarly, it is not mandatory to drop other operator classes since
we are adding all the operators into fixeddecimal_ops family so
other operator classes will mostly be unused and we will drop them
conditionally if no dependencies found for them.
Additionally, replaced DROP VIEW sys.sysindexes statement with
rename and recreate statements to fix possible upgrade failure.
Manually tested with test case added with original change (BABEL-2089.sql)
as well as with primary key's dependency on fixeddecimal_ops class.
Task: BABEL-3544

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>
Co-authored-by: Rishabh Tanwar <ritanwar@amazon.com>

timchang514 committed:
[BABEL-3156] Support UPDATE, DELETE for sp_describe_undeclared_parameters (#690)

pratikzode committed:
Remove babelfishpg_tsql.enable_ownership_structure. (#707)

* “Remove babelfishpg_tsql.enable_ownership_structure.
Previously, The GUC babelfishpg_tsql.enable_ownership_structure no longer serves a purpose as Babelfish cannot run without the ownership structure. Now have removed ownership structure.
Task: BABEL-3154
Signed-off-by:Pratik Zode<pzode@amazon.com>“

Signed-off-by: pratikzode <pratik.zode@gmail.com>

* Update on Remove babelfishpg_tsql.enable_ownership_structure
Had removed the Enable Babelfish Ownership Structure from gun.c and also removed this function's declaration in guc.h as well

Task: BABEL-3154
Signed-off-by: pratikzode <pratik.zode@gmail.com>

* Update on guc.c file

Task: BABEL-3154
Signed-off-by: pratikzode <pratik.zode@gmail.com>

Signed-off-by: pratikzode <pratik.zode@gmail.com>

jsudrik reviewed (commented):
Thanks for adding this. 

Can we please leave one line gap under every title so that it becomes more obvious that these are the sections to be filled?

Will the github actions fail the attempt to merge if the template is not present?

shalinilohia50 committed:
Add metadata integrity check for change in sys.babelfish_authid_user_ext

Add a rule for metadata integrity check to check if all the newly
created databases have a guest user.

Task: BABEL-3600
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

ColinKYuen committed:
Cancel previous GitHub Action workflows on same branch with new commits (#667)

Description
Changes
Cancels previous GH Action workflows in the same branch when there is a new commit to it.
Pushes to PR only run the actions once rather than twice (Push & PR), clarification: this is more for forked repositories with write access.

Benefits
Reduce wait time for actions to run if there are multiple commits to the same branch
Reduces costs of using Github Action minutes

Note
It runs on whenever a PR is created / has a new push (to the same branch)
It runs whenever a branch with the following prefix BABEL_* is pushed to, this is to have it still run on default branch.

Signed-off-by: Colin Yuen yuenhcol@amazon.com

Commit:
Add common psqlODBC tests file (#716)

Add common psqlODBC tests file. Also add sys.datetime & sys.bigint data type tests using the common test functions in this file.

TASK: BABELFISH-584

Signed-off-by: Avneet Toor (avneett@bitquilltech.com)

shalinilohia50 committed:
Incorrect result when selecting from tempdb.sys.objects

Task: BABEL-3211
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

anujsarda3 committed:
Add JSON_MODIFY() (#635)

This function is used to update the value of the property in JSON string. Test cases are also added for the implementation.

Signed-off-by: Anuj Sarda sardanuj@amazon.com
Task: BABEL-937

rossd-bitquill committed:
Add explicit casts for various numeric datatypes to SMALLDATETIME (#694)

This commit adds support for the following numeric types being cast to
SMALLDATETIME:
bit, numeric, decimal, float, real, int, bigint, smallint, tinyint, money, smallmoney

Task: BABELFISH-536
Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

jusjosj committed:
Add allow hints for delete update and insert in stored procs (#696)

Previously, using tsql style hints on non select statements inside of
stored procs was not supported.  This commit allows for hints on update
delete and insert statements inside of stored procedures.

This is enabled by calling the process_execsql_remove_unsupported_tokens
inside of exitdml_statement in the select statement mutator. This will
properly sanitize the query strings an extract the hints.

Further process_execsql_remove_unsupported_tokens was updated to
consider the expression mutators context as the base context when
rewritting queries.  For normal statements this was the same as the old
behavior since the offset will be 0.  This is required for stored procs
as the former behavior would not account for the "CREATE PROCEDURE <name>
AS" in the expression thus causing a rewrite at the wrong location

Task: BABEL-3592
Signed-off-by: Justin Jossick jusjosj@amazon.com

Suthapalli-Ramya-satya-vasavi-srija committed:
Moved babelfish related content from rdsutils extension to tdsutils (#639)

* Moved babelfish related content from rdsutils extension to tdsutils

Previously, some babelfish code logic related to dropping and renaming
has been put in rdsutils extension. Now, it has been moved to tdsutils part.

Task: BABEL-3457
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

* Fix github workflows to point to correct branches (#641)

Github actions should point to the stable branches and the latest dev branches for running the upgrade tests.

* Add 'w' unit to four date functions (#501)

* Add 'w' unit to four date funtions

Add 'w' unit to DAREPART(), DATENAME(), DATEDIFF(), DATEADD() functions
by add 'w' as a keywork.

Task: BABEL_3358
Signed-off-by: Zhenye Li <zhenye@amazon.com>

* Correct dateiff_arg

Signed-off-by: Zhenye Li <zhenye@amazon.com>

* Correct datepart argments

Signed-off-by: Zhenye Li <zhenye@amazon.com>

* retrigger checks

* Update gram-tsql-rule.y

* Delete tests from upgrade schedule files

Delete tests from upgrade schedule files and update
expected_dependency.out

Signed-off-by: Zhenye Li <zhenye@amazon.com>

Signed-off-by: Zhenye Li <zhenye@amazon.com>
Co-authored-by: Zhenye Li <zhenye@amazon.com>

* Expand explain for PRINT, THROW, USE and RAISERROR (#561)

This commit extends the functionality of babelfish_showplan_all to include print, throw, raiserror, and use
This is largely done in a manner consistent with the other explain functions existing throughout the codebase.

Task: BABEL-2902
Signed-off-by: Justin Jossick <jusjosj@amazon.com>

* Change github actions to point to correct engine dev branch (#649)

Authored-by: Nirmit Shah <nirmisha@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

* Fix table hint parsing (#637)

Previously, Babelfish would sometimes incorrectly parse a query that used a table alias with column definitions as a table hint and raise an error. This commit fixes the issue by having the parser only accept a specific defined list of keywords as possible table hints.

N.B. Babelfish will report incorrect table hint syntax as generic syntax errors instead of specifically table hint errors as SQL Server does. This commit does not attempt to mimic this behavior, as that would require a more extensive modification to the antlr parser (most likely requiring a new grammar rule to be added).

Task: BABEL-2805
Signed-off-by: Jason Teng jasonten@amazon.com

* Add JDBC tests for CONVERT()

An issue BABEL-2054 was reported where CONVERT(datetime, string) failed
intermittently. However, this issue is not reproducible and this commit
adds JDBC tests to validate the scenario mentioned in the JIRA

Task: BABEL-2054

Signed-off-by: Amith Kamath <amithkam@amazon.com>

* Login who owns database should always be dbo also without sysadmin role

In SQL Server, the login that creates a database is automatically the
database owner. This means that the login can always access the
databases they create.

However, previously in babelfish, if a login is stripped of the sysadmin
role after creating a database, and the login does not have a
user associated with the database, then the login can no longer access
the database. Now, when a login attempts to access and there
is not an associated user, the login will be considered the dbo and
allowed access.

Task: BABEL-3548

Signed-off-by: Walt Boettge <wboettge@amazon.com>

* Included test files related to alter/dropping babelfish roles and modified failing tests. Addes get_rolespec_name_internal in tdsutils as get_rolespec_name function is different in open source and internal engine codes.

* Add tests related to alter/drop babelfish created roles

* Include header file for table_open function

* Include required header file for function call

Signed-off-by: vasavi suthapalli <svasusri@amazon.com>
Signed-off-by: Zhenye Li <zhenye@amazon.com>
Signed-off-by: Justin Jossick <jusjosj@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>
Signed-off-by: Jason Teng jasonten@amazon.com
Signed-off-by: Amith Kamath <amithkam@amazon.com>
Signed-off-by: Walt Boettge <wboettge@amazon.com>
Co-authored-by: Nirmit Shah <44708230+shah-nirmit@users.noreply.github.com>
Co-authored-by: ZhenyeLee <73271404+ZhenyeLee@users.noreply.github.com>
Co-authored-by: Zhenye Li <zhenye@amazon.com>
Co-authored-by: jusjosj <96209028+jusjosj@users.noreply.github.com>
Co-authored-by: Jason Teng <akariass@gmail.com>
Co-authored-by: AmithKamath <kamath.amith64@gmail.com>
Co-authored-by: Walt Boettge <wboettge@amazon.com>

faviansamatha committed:
Improve sp_sproc_columns performance (#708)

* Improve sp_sproc_columns performance

Previously, sp_sproc_columns had major performance issue due to some
convoluted joins. This commit fixes those issues by reworking the query.
Furthermore, this commit also allows sp_sproc_columnst to detect
Table-valued parameters (TVP) along with other minor improvements such
as showing correct COLUMN_TYPE value for output parameters in stored
procedures, and corrected fields for the return value of
Table_Returned_values.

Task:BABEL-3555, BABEL-3595

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Modified dep-vu-verify

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Addressed PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* cleanup and comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* FFixed several int types by prepending sys.

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Addressed some PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Addressed PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Fixed the NULLABLE column for stored procedure return values

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Addressed PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* Fixed indents and formatting

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

* removed unecessary where clause and replaced information_schema_tsql with regular information_schema

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

ColinKYuen committed:
PSQLODBC Datatype Test for BYTEA (#718)

Adds ODBC Breakthrough Datatype Test for BYTEA
Create Table, Insert valid values, Update table with valid values, Comparison Operators, Table View, Unique Constraint, Composite Key

Task: BABELFISH-580
Signed-off-by: Colin Yuen yuenhcol@amazon.com

2jungkook committed:
Avoid duplicate OBJECT_IDs after MVU (#693)

PG allows duplicated OIDs in different catalog tables. However, in Babelfish, we use PG's OID as OBJECT_ID in all system catalog views and it should be unique within a logical database (https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-catalog-views/sys-all-objects-transact-sql.md). Duplicate OIDs can break SQL server tools, e.g., SSMS, as well as customer applications.

In PG/Babelfish, duplicate OIDs can occur in the following two cases:
- After MVU (Major Version Upgrade)
- After hitting OID wraparound

This commit is to prevent duplicate OIDs during MVU and consists of two parts:

First of all, from the old cluster, it finds the max OID in the 5 catalog tables: pg_extension, pg_authid, pg_enum, pg_class and pg_type. And then pass the max OID to the new cluster as a form of GUC, babelfishpg_tsql.dump_restore_min_oid. In the new cluster, set the nextOid greater than the max OID from the old cluster using GetNewObjectId_hook.

Secondly, in pg_upgrade.c, it skips the step of transferring the next OID from the old cluster to the new cluster.

Also note that the type of babelfishpg_tsql.dump_restore_min_oid is defined by DefineCustomStringVariable() to support UINT_MAX.

Also, this commit contains minor improvements on the pg-upgrade Github action and dev-tools.sh:
- pg-upgrade Github action: add retain option to see pg_upgrade log even if pg_upgrade succeeds and jdbc tests fails.
- dev-tools.sh: developers can choose the migration mode, e.g., ./dev-tools.sh test input multi-db.

Engine PR: babelfish-for-postgresql/postgresql_modified_for_babelfish#54

Task: BABEL-3613
Signed-off-by: Jungkook Lee jungkook@amazon.com

ColinKYuen committed:
PSQLODBC Datatype Test for TIME (#720)

Adds ODBC Breakthrough Datatype Test for TIME:
Create Table
Insert valid values
Insert invalid values
Update table with valid values
Update table with invalid values
Comparison Operators
Table View
Unique Constraint
Composite Key

Task: BABELFISH-577
Signed-off-by: Colin Yuen yuenhcol@amazon.com

faviansamatha committed:
Fixed identifying IF and TF in sys.all_objects and OBJECTPROPERTY (#541)

Previously, sys.all_objects mistakenly identified inline table-valued functions (IF), and table-valued functions (TF) as scalar functions (FN). This commit fixes that issue. In addition, this commit fixes the issue for the following properties in OBJECTPROPERTY: IsScalarFunction, IsTableFunction, IsInlineFunction. This commit also fixes misidentifying table functions as scalar functions in SSMS.

Tasks: BABEL-3369, BABEL-3370, BABELFISH-511

Signed-off-by: Favian (Ian) Samatha ians@bitquilltech.com

Commit:
Add "USE master" statement in Test-xp_instance_regread test file (#730)

Ensure tests for xp_instance_regread will always run in master database

Signed-off-by: Avneet Toor avneett@bitquilltech.com

Suthapalli-Ramya-satya-vasavi-srija committed:
Support sp_addrole, sp_addrolemember, sp_droprole, sp_droprolemember

This commit supports sp_addrole, sp_addrolemember, sp_droprole, sp_droprolemember
- sp_addrole: Creates a new database role in the current database.
- sp_droprole: Removes a database role from the current database.
- sp_addrolemember: Adds a database user, database role, Windows login, or Windows group to a database role in the current database.
- sp_droprolemember: Removes a security account from a SQL Server role in the current database.

Task: BABEL-3087
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

raydhkim committed:
Fix: CREATE USER should not be allowed for the login owning the database

In SQL Server, a login that owns a database is not allowed to create a DB user in that owning database. However, Babelfish allowed this behavior, not throwing an appropriate error when a DB-owning login tries to create a DB user in the database. Now, it follows the behavior of SQL Server, not allowing db-owning login to create a user in the database, by checking if current login name and current database’s owner name match.

Task: BABEL-3549
Signed-off-by: Ray Kim <raydhkim@amazon.com>

Suthapalli-Ramya-satya-vasavi-srija committed:
Disallow ALTER ROLE in PG for Babelfish-created roles

Previously, for a PG user to break a Babelfish instance by using
ALTER ROLE to block logins, change role membership, change passwords,
and other changes that would have a dramatic effect on Babelfish
functionality. Now, ALTER ROLE been disabled for Babelfish-created PG
roles.

Task: Babel-2976
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

pratikzode committed:
Add a GUC to disable metadata inconsistency check function. (#723)

* Add a GUC to disable metadata inconsistency check function.
For Babelfish instances, CP would stop the upgrade workflow when there are some metadata inconsistencies detected.
To prevent false alarm, we should have a GUC to disable the metadata check function babelfish_inconsistent_metadata.
Now we have added babelfish_inconsistent_metadata in the guc file

Task: BABEL-2967
Signed-off-by: pratikzode <pratik.zode@gmail.com>

ColinKYuen committed:
PSQLODBC Datatype Test for sys.bit (#724)

Adds ODBC Breakthrough Datatype Test for sys.bit

Task: BABELFISH-562
Signed-off-by: Colin Yuen yuenhcol@amazon.com

Suthapalli-Ramya-satya-vasavi-srija committed:
Add default case for tdsutils_processutility

Previously, while building the babelfish_extensions package there
are many warnings related to not including test case. Resolved it
by adding default case in function tdsutils_processutility.

Task: N/A
Signed-off-by: vasavi suthapalli <svasusri@amazon.com>

Sairakan committed:
Fix behavior of CROSS JOIN on parenthesized table expressions (#706)

When APPLY was originally implemented, a subtle issue was introduced
whereby queries that invoked CROSS JOIN with a parenthesized table expression
on the right side would raise a syntax error. This commit reverts that change
and instead implements APPLY via an ANTLR rewrite.

Task: BABEL-2805
Signed-off-by: Jason Teng <jasonten@amazon.com>

ColinKYuen committed:
Revert "Cancel previous GitHub Action workflows on same branch with new commits (#667)" (#738)

Revert GitHub Actions Cancellation using Concurrency due to tests being cancelled in the main branch
This reverts commit 18eabd00a4483c6662b1d8489453fdd4c33237c7.

Signed-off-by: Colin Yuen yuenhcol@amazon.com

Commit:
psqlODBC tests for sys.datetime2, sys.datetimeoffset, sys.smalldatetime, sys.real (#725)

Add psqlODBC data type tests for sys.datetime2, sys.datetimeoffset, sys.smalldatetime, & sys.real.

Signed-off-by: Avneet Toor avneett@bitquilltech.com

Issues Resolved
BABELFISH-547, BABELFISH-559, BABELFISH-570, BABELFISH-571

ColinKYuen committed:
PSQLODBC Datatype Test for DATE (#726)

Adds ODBC Datatype Test for DATE

Task: BABELFISH-576
Signed-off-by: Colin Yuen yuenhcol@amazon.com

rishabhtanwar29 committed:
Fix T-SQL coercion hash table (#713)

Description
Commit fe0a0a7 implemented a fix to load datatype coercion hash table
on demand along with a logic to keep initializing the hash table on every
lookup until all the hash table entries are initialized completely.
But later it was discovered that some of the entries of static array (which we
use to fill the hash table with) are not correctly written so they never get
filled into the hash table and as a result we keep initializing the hash table
on every lookup.

This commit addresses the issue by fixing all incorrect entries in the static
array as well as adds additional logic for searching a cast function.

basasairohan committed:
Added missing condition in sys.all_objects view definition (#736)

* Added missing condition in sys.all_objects view definition

* Added tests in BABEL-3613 upgrade script

* Modified expected output files

* Removed duplicate definition of sys.all_objects in upgrade script.

timchang514 committed:
Various bugs with FORMAT() (#722)

Fixing various bugs with FORMAT():

- Spaces in the currency and datetimeformat tables in format.h were using &nbsp instead of regular spaces, so would appear weirdly in output files viewed in certain encodings.
- 'yyy' was returning the same output as 'yyyy'.
- 't' and 'tt' in sequence would only output every other one.
- smallmoney_testing table was using the money datatype.
- For XML was broken, as it was constructing a query to use 'format'. We specify the schema to be 'pg_catalog' to avoid issues.

Task: BABEL-3598
Signed-off-by: Tim Chang changti@amazon.com

ColinKYuen committed:
PSQLODBC Datatype Test for sys.money (#731)

Adds ODBC Datatype Test for sys.money

Task: BABELFISH-549
Signed-off-by: Colin Yuen yuenhcol@amazon.com

ColinKYuen committed:
PSQLODBC Datatype Test for sys.uniqueidentifier (#734)

Adds ODBC Datatype Test for sys.uniqueidentifier

Task: BABELFISH-569
Signed-off-by: Colin Yuen yuenhcol@amazon.com

ColinKYuen committed:
PSQLODBC Datatype Test for sys.binary & sys.varbinary (#742)

Adds ODBC Datatype Test for sys.binary & sys.varbinary
Added edge cases for timezone helper in sys.datetimeoffset
Removed outdated comments in sys.datetimeoutset
Added driver check to sys.real's teardown

Task: BABELFISH-568 & BABELFISH-572
Signed-off-by: Colin Yuen yuenhcol@amazon.com

ZhenyeLee committed:
Support duplicate trigger name in different databases (#680)

Previously, duplicate trigger name was not allowed in different databases.
But in tsql, different databases can have the same trigger name. Here, we 
check if a trigger with the same name is in current database. If not, then the 
trigger can be created. And during drop trigger, we modify the pre check 
trigger in parsing stage to not throw error, and find the correct trigger to 
drop with trigger and database name. 

Task:BABEL-3117
Signed-off-by: Zhenye Li <zhenye@amazon.com>

2jungkook committed:
Add babelfishpg_tsql.enable_pg_hint GUC (#744)

This commit introduces babelfishpg_tsql.enable_pg_hint GUC that can load pg_hint_plan on-demand. Also it revises dev-tools.sh and Github actions to build pg_hint_plan by default. However, users still can build Babelfish without pg_hint_plan. In addition to that, it moves T-SQL hint-related tests into the input/pg_hint_plan from pg_hint_plan directory.

Task: BABEL-3513
Signed-off-by: Jungkook Lee <jungkook@amazon.com>

Commit:
Add psqlODBC data type tests for sys.smallmoney (#747)

Add psqlODBC data type tests for sys.smallmoney

These are the created tests

Table Creation
Insertion_Success
Insertion_Fail
Update_Success
Update_Fail
View Creation
Table_Single_Primary_Keys
Table_Composite_Primary_Keys
Table_Unique_Constraint
Comparison_Operators
Arithmetic_Operators
Comparison_Functions
Also added missing const keyword in front of variables in various data type test files.

Signed-off-by: Avneet Toor avneett@bitquilltech.com

jusjosj committed:
Make substring null behavior congruent with tsql (#678)

This commit addresses a difference in the way that substring handled null arguments between babelfish and tsql. Prior to this change, babelfish would raise an error on any null input to the substring function regardless of whether or not this is a null constant or another data type with the value of null. After this change, babelfish will only raise an error on values which are considered to be null constants restoring paraity to tsql.

Task:BABEL-3347
Signed-off-by: Justin Jossick jusjosj@amazon.com

shalinilohia50 committed:
sp_helpuser 'guest' shows SID = NULL (#751)

The procedure sp_helpuser should show SID value as 0x00 for the guest user but currently it is displaying NULL. With this change, sp_helpuser 'guest' shows SID value as 0x00.

HarshLunagariya committed:
Support TSQL ownership chaining within same DB in procedure

This commit supports forming ownership chain while executing the
procedure such that when referenced object(in any particular statement)
and procedure has same owner then invoker will be able to execute that
statement even if it doesn't have explicit permissions on referenced
object.
To implement above functionality we've implemented hook which
- check whether current statement execution is inside procedure or not by
looking at the PLtsql_exec_state stack.
- If it is inside procedure then it compares the referenced object and
procedure's owner.
- If both are same then this hook bypasses the permission check and allows
execution of that statement.

Task: BABEL-3014
Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

faviansamatha committed:
Add psqlODBC tests for sys.sql_variant (#743)

This commit adds the sys.sql_variant tests for psqlODBC. It adds the
following test cases:

-Create Table
-Insert valid values
-Update table with valid values
-Comparison Operators
-Table View
-Unique Constraint
-Composite Key
-Primary Key

Task: BABELFISH-592
Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

Commit:
Add testStringFunctions common function (#753)

Add a testStringFunctions for psqlODBC tests. This function verifies the expected results for various string functions (LOWER, UPPER, TRIM, etc.).

Signed-off-by: Avneet Toor avneett@bitquilltech.com

raydhkim committed:
Fix newly added error message for unsupported HIERARCHYID, GEOGRAPHY, GEOMETRY datatypes

Previously, the newly added error messages for HierarchyID, Geography,
Geometry had a reference to escape_hatch_rowversion, which is not
correct. Now, the error messages accurately refer to each escape_hatches
accordingly.

Task: BABEL-2517

Signed-off-by: Ray Kim <raydhkim@amazon.com>

faviansamatha committed:
Add psqlODBC tests for sys.decimal and numeric (#754)

This commit adds the sys.decimal and numeric tests for psqlODBC. It adds
the following test cases:

Table Creation
Failed Table Creation
Insert Valid Values
Insert Invalid Values
Update Valid Values
Update Invalid Values
Table View
Primary Key
Unique Constraint
Composite Keys
Arithmetic Operators
Comparison Operators

Task: BABELFISH-554, 592
Signed-off-by: Favian (Ian) Samatha ians@bitquilltech.com

ColinKYuen committed:
PSQLODBC Datatype Test for CHAR, TEXT & sys.nvarchar (#748)

Added psqlODBC tests for char, text, & sys.nvarchar

Task: BABELFISH-552, BABELFISH-579, & BABELFISH-575
Signed-off-by: Colin Yuen yuenhcol@amazon.com

HackonToby committed:
Add psqlODBC test for sys.bpchar (#740)

Add test case for sys.bpchar

Task: BABELFISH-548
Signed-off-by: Toby Cao tobyc@bitquilltech.com

HackonToby committed:
Add psqlODBC test for sys.varchar (#739)

Add test case for sys.varchar

TASK: BABELFISH-551
Signed-off-by: Toby Cao tobyc@bitquilltech.com

HackonToby committed:
Add psqlODBC test for sys.nchar (#750)

Add test case for sys.nchar

TASK: BABELFISH-550

Signed-off-by: Toby Cao tobyc@bitquilltech.com

HackonToby committed:
Add psqlODBC test for sys.ntext (#732)

Add test case for sys.ntext.
Added required const to sys.nchar, sys.varchar, sys.bpchar.

TASK: BABELFISH-555
Signed-off-by: Toby Cao tobyc@bitquilltech.com

KushaalShroff committed:
Performance optimisation for INSERT BULK

Initially Insert Bulk was implemented using SPI_execute_with_args, which practically were single inserts in the executor level. This made database migration slow for our users, with this task we have moved on from SPI to use PG’s multi insert functionality. PG’s multi insert feature which is much faster since we can write just a single WAL for all of the inserts at the end and we need to lock/unlock page only once. Moreover, there is no need to parse/plan/optimise the query (When compared to initial implementation) thus bypassing directly to inserting multiple tuples into the table at once. A speedup of 8x on average was achieved when compared to the initial implementation.

Task: BABEL-3503
Authored-by: Kushaal Shroff <kushaal@amazon.com>
Signed-off-by: Kushaal Shroff <kushaal@amazon.com>

bsatarupa committed:
Arithmetic overflow error for NUMERIC datatype (#701)

Problem description
Existing bbf code throws overflow error for multiplication or division of numeric operands whenever the resultant calculated precision > TDS_MAX_NUMERIC_PRECISION(i.e., 38) and resultant calculated integral part(i.e., precision - scale) > 32 and resultant calculated scale > 6

Ideally, the above scenario may and may not cause overflow and that decision is made later based on the actual numeric datalength (which is already handled as part of resolving BABEL-3179).

We have fixed the above mentioned issue, by rounding scale to 6 and precision to TDS_MAX_NUM_PRECISION.(mimicking sql server behaviour: https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/data-types/precision-scale-and-length-transact-sql.md) and deferred the overflow checking decision to a later phase when integral part of actual data can't be accommodated in 32 digits (i.e., entire numeric result can't be accommodated in TDS_MAX_NUM_PRECISION digits since scale is already set to 6).

bsatarupa committed:
[BABEL-2795] DATEDIFF() returns incorrect result (#699)

Problem description
As per current approach, while calculating day difference count between two dates bbf returns the number of full days between the dates using PG's internal date_part function, date_part takes hh:mm:ss.msec into consideration.
Fix includes, calculating daydiff solely from dates, without considering timestamp (i.e., hh:mm:ss.msec). Same goes for hourdiff. Made changes accordingly.

Upgrade
This commit includes a new sql function that calculates number of days elapsed from 01/01/0001 for any given date; which is used to calculate daydiff between two input dates.

basasairohan committed:
Add to the JDBC test framework the ability to set a performance SLA per file (#700)

Description
Currently the tests developed on Babelfish do not have performance SLAs. This change adds SLAs to the JDBC test framework.
Added a new parameter in config.txt to specify the default SLA. This will be considered as default timeout unless it is explicitly mentioned in the test file.
One can specify the SLA value in milli seconds for a specific test file by adding a comment at the starting of the test file mentioning the SLA value. Ex: -- SLA 10000

kuntalghosh committed:
Improve identity value insert performance with IDENTITY_INSERT=on (#728)

With IDENTITY_INSERT=on in a table, users are allowed to insert values in identity columns. We also need to update the last used identity for the table(seed) so that subsequent insertion with default identity can start from the same seed. As per the current approach, after inserting one row, we're doing a full scan of the table to find out the min/max identity value and updating these variables. This creates a huge performance bottleneck for data insertion and the performance degrades significantly as the number of rows increases in the table.

In T-SQL, the last used identity of a table is non-transactional. That means, it always increases (or decreases) irrespective of the underlying transaction commits/rollbacks. To implement this, we don't have to scan the entire table. We can just calculate the max(current seed, value to be inserted) and update the seed under the same lock.

Task: BABEL-3506
Signed-off-by: Kuntal Ghosh kuntalgh@amazon.com

shah-nirmit committed:
Fixed the default encoding in case of collation lookup failed (#760)

Previously We hardcoded the Server encoding to be used for string datatypes when the collation Lookup failed (Ex PG Default(100), "C"(950) collations). This leads to improper cast of special chars. (Ex pound sign) to varchar datatype when called in PG_Dialect .With this change we set the encoding based on default server collation in case collation could not be found.

Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

Jakeowen1 committed:
TRY_CONVERT to datetime2 causes NULL return (#712)

Currrently, Babelfish returns NULL when a user calls TRY_CONVERT to convert a datatype to a datetime2. This commit enables TRY_CONVERT to behave correctly with the datetime2 datatype. This commit also updates tests from BABEL-3486 with the correct expected output for TRY_CONVERT calls with datetime2.

Task: BABEL-3614
Signed-off-by: Jake Owen <owjco@amazon.com>

Commit:
psqlODBC data type tests for int, tinyint, and float (#741)

Add data type tests for int, tinyint, and float.
Added a conversion_functions_common file that contains all common C++ type conversions required by some psqlODBC tests
Capitalized const variables.
Added comparison operator test to sys.real

Signed-off-by: Avneet Toor avneett@bitquilltech.com

TASKS: BABELFISH-546, BABELFISH-553, BABELFISH-560

jusjosj committed:
Add planning and execution times to babelfish_statistics profile (#717)

This commit changes babelfish_statistics profile in two ways. First it enables the summary and timing gucs by default to match postgres. Second it implements the summary timings. This is done through adding variables in estate that are set when the appropriate hook functions are called.

For planning, a new hook was introduced as many babelfish code paths will invoke the planner from a SPI_exec function. This makes mirroring postgres which during its explain analyze flow only calls through to the pg_plan_query function. By adding this hook, we are able to time standard_planner which is the majority of the work done in pg_plan_query.

For execution time, we can simply time the existing executor hooks. The closing calculation is located in append_explain_info as to match postgres, we must only stop the timer after we have written out the rest of the metrics.

Task: BABEL-3338
Signed-off-by: Justin Jossick jusjosj@amazon.com

HackonToby committed:
Add psqlODBC test for sys.sysname (#729)

Add test cases for sys.sysname

TASK: BABELFISH-556

Signed-off-by: Toby Cao tobyc@bitquilltech.com

AvantikaDG committed:
Upload logs when extension install fails

In jdbc tests the postgres logs are uploaded only if the tests fail.
In this commit we upload the logs even when there is a failure in the
extension installation step, for easier debugging.

Task: BABEL-3626
Signed-off-by: Avantika Dasgupta <davantik@amazon.com>

AvantikaDG committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into davantik-dev
